### PR TITLE
Adding path derivation

### PR DESCRIPTION
**Pull request by runn1ng**
Now this might be a bit controversial, but I found something like that to be really useful.

What will be controversial, I think:

* the usage of strings as indexes 
   * yes, it's not ideal, but adding an apostrophe is a way to easily say "this is hardened" and this notation is common, and you cannot do that otherwise
* the usage of descriptions as paths in the tests
   * again, not ideal, but I didn't want to repeat the description as path again if it's already there

### CODE REVIEW

**Comment by dcousens** [comment]
ping @jprichardson,  merge here,  finalize API,  then extract to `BIP32`?

File: src/hdnode.js [hunk 1]
```diff
@@ -281,6 +281,24 @@ HDNode.prototype.deriveHardened = function (index) {
   return this.derive(index + HDNode.HIGHEST_BIT)
 }
 
+// number + ' (e.g. 234') as a string -> hardened
+HDNode.prototype.derivePath = function (path) {
```
Comment (dcousens) [inline src/hdnode.js]: Any reason why this just doesn't take a `string` as input?

**Comment by dcousens** [comment]
IMHO,  I think  `node.derivePath('m/0'/1/0')` would be a better API,  aka a `String` argument.

I don't think there is much value in adding an `Array` based path API.  You can trivially do that using `derive` and `forEach` already.

**Comment by dcousens** [comment]
We also intend to break out `BIP32` to its own module,  probably encompassing `HDNode` standalone with other utilities for making life easier when dealing with them.

See https://github.com/bitcoinjs/bitcoinjs-lib/issues/508#issuecomment-176188789

**Comment by runn1ng** [comment]
ad array based API:

My original thinking for using Array was "What if someone puts there an array that is *not* a path, or that is wrongly formatted? Do I have to trim spaces around the slashes? Does it have to start with 'm/' or not? What if the node is not a master and is a sub-node, the 'm/' doesn't make sense", and that by using Array I just go around all these problems.

Then I also thought about the function accepting both array and string, but that would be really ugly.

**Comment by jprichardson** [comment]
> ping @jprichardson, merge here, then extract to BIP32?

@dcousens Yep, I don't see any harm in adding it at the moment. Agreed on not needing the array API.

**Comment by runn1ng** [comment]
how exactly would you do the string-based API? Should it start with 'm/', or should it be "relative"? So for example,` master.derivePath("0'/1/0'/1") `(which would equal to `master.derivePath("0'/1").derivePath("0'/1")`)

**Comment by dcousens** [comment]
I think `0'/1/0` is a suitable enough API.
We could easily regex,  optionally skipping any prefixes until the first number is reached?

**Comment by runn1ng** [comment]
By "we", you mean the code in the library, or the user of the API?

**Comment by dcousens** [comment]
We, the developers of this library.
My suggestion was that we could make the `m/` optional,  but,  ambiguous API's suck,  so,  not sure.

**Comment by runn1ng** [comment]
>ambiguous API's suck

Exactly. But then, I don't know how to make it unambiguous and non-sucky.

And I don't also want to totally abandon it either. 😓

**Comment by dcousens** [comment]
@runn1ng I think if an `m/` is present,  we check that there is no parent (aka, this is a `master` node).
If it isn't present,  proceed on with derivation.

Thoughts?

**Comment by runn1ng** [comment]
Sorry for the delay. OK, that sounds reasonable

**Comment by runn1ng** [comment]
Done :)

File: src/hdnode.js [hunk 2]
```diff
@@ -291,6 +291,30 @@ HDNode.prototype.isNeutered = function () {
...
+  typeforce(types.String, path)
+
+  var splitPath = path.split('/')
+  if (splitPath[0] === 'm') {
+    if (!this.parentFingerprint) {
+      splitPath = splitPath.slice(1)
+    } else {
+      throw new Error('Not a master node')
+    }
```
Comment (dcousens) [inline src/hdnode.js]:
``` javascript
if (this.parentFingerprint) throw new Error('Not a master node')
```

Comment (runn1ng) [inline src/hdnode.js]: I am not sure standard allow if block without curly braces :) but yeah I got you. Ok, will do tomorrow

**Comment by runn1ng** [comment]
I have done some cosmetic changes and rebased. The failing check has nothing to do with my code.

File: src/hdnode.js (hunk 2 above)
Comment (runn1ng) [inline src/hdnode.js]: done

**Comment by dcousens** [comment]
Sounds good, will review asap :)

**Comment by fanatid** [comment]
@runn1ng why unary plus instead `praseInt`?
with unary plus string `/` equals `0/0` because `+''` gives `0`, with `parseInt` string `/` throw error.
```js
return splitPath.reduce(function (prevHd, indexStr) {
  if (indexStr.slice(-1) !== "'") return prevHd.derive(parseInt(indexStr, 10))
  return prevHd.deriveHardened(parseInt(indexStr.slice(0, -1), 10))
}, this)
```

**Comment by runn1ng** [comment]
Good catch.

**Comment by runn1ng** [comment]
Just thinking.... if you put in path with decimal points ("m/0.5/1") by some mistake, using "parseInt" will silently truncate it instead of throwing errors.

But I am not sure that is worth caring about.

**Comment by fanatid** [comment]
As option, you can check that result of `parseInt` is equal original string, something llike this:
```js
return splitPath.reduce(function (prevHd, indexStr) {
  var derive = prevHd.derive
  if (indexStr.slice(-1) === "'") {
    derive = prevHd.deriveHardened
    indexStr = indexStr.slice(0, -1)
  }

  var index = parseInt(indexStr, 10)
  if (index.toString() !== indexStr) throw new Error(...)

  return derive.call(prevHd, index)
}, this)
```

**Comment by runn1ng** [comment]
I am also thinking if things like `m/   2   /   0'  /` should be allowed. `parseInt` will munch it just fine.

I will probably do what you suggest.

Alternatively, I can just test the index with regexp if it's only numbers.

**Comment by dcousens** [comment]
`/([m]\/)?([0-9]+[']?\/)*([0-9]+[']?)/`

**Comment by runn1ng** [comment]
Well, I wanted to test indices one by one after splitting, which is more readable and easily debuggable, but it's true that testing the parameter at the start of a function is more in line with the rest of the code.

**Comment by runn1ng** [comment]
I have added path validation to typeforce types.

File: src/types.js [hunk 3]
```diff
@@ -26,6 +26,11 @@ function UInt53 (value) {
     Math.floor(value) === value
 }
 
+function Path (value) {
```
Comment (dcousens) [inline src/types.js]: `BIP32Path`?  *shrug*

It's private anyway,  so I'm not worried too much.

File: test/fixtures/hdnode.json [hunk 4]
```diff
@@ -134,7 +134,7 @@
           "address": "1BxrAr2pHpeBheusmd6fHDP2tSLAUa3qsW"
         },
         {
-          "description": "m/0/2147483647'/1/2147483646",
+          "description": "m/0/2147483647'/1/2147483646'",
```
Comment (dcousens) [inline test/fixtures/hdnode.json]: Hah.  Good catch.

File: test/hdnode.js [hunk 5]
```diff
@@ -285,6 +286,42 @@ describe('HDNode', function () {
           verifyVector(hd, c, i + 1)
         })
       })
+
+      // testing deriving path from master
+      f.children.forEach(function (c) {
+        it(c.description + ' from ' + f.master.fingerprint + ' by path', function () {
+          var path = c.description
+          var child = master.derivePath(path)
```
Comment (dcousens) [inline test/hdnode.js]: Nice :)

File: test/hdnode.js [hunk 6]
```diff
@@ -368,6 +405,30 @@ describe('HDNode', function () {
...
+      }, /Expected Path/)
+      assert.throws(function () {
+        master.derivePath([2, 3, 4])
+      }, /Expected Path/)
+      assert.throws(function () {
+        master.derivePath('/')
+      }, /Expected Path/)
+      assert.throws(function () {
+        master.derivePath('m/m/123')
+      }, /Expected Path/)
```
Comment (dcousens) [inline test/hdnode.js]: Its not great,   but,  its not terrible either.  ACK for now with room for improvement haha

Comment (fanatid) [inline test/hdnode.js]:
```js
[
  'foo',
  '',
  2,
  ...
].forEach(function (path) {
  assert.throws(function () { master.derivePath(path) },  /Expected Path/)
})
```

File: src/types.js (hunk 3 above)
Comment (fanatid) [inline src/types.js]: +1 for `BIP32Path`

File: test/hdnode.js (hunk 6 above)
Comment (dcousens) [inline test/hdnode.js]: And put those negative fixtures in the json fixtures under `invalid: { paths: [...] }`

Comment (dcousens) [inline test/hdnode.js]: See https://github.com/bitcoinjs/bitcoinjs-lib/blob/master/test/fixtures/hdnode.json#L195 for examples

Comment (dcousens) [inline test/hdnode.js]: @runn1ng if you're time is running short on this (not that there has been any hint of that),  I'll see if I can submit a PR to this branch (on your repo) to help out :)

Comment (runn1ng) [inline test/hdnode.js]: All right, I can do that

File: src/types.js (hunk 3 above)
Comment (runn1ng) [inline src/types.js]: Ok

**Comment by runn1ng** [comment]
OK, feedback was put in

**Comment by fanatid** [comment]
ACK

**Comment by dcousens** [comment]
ACK :+1:

**Comment by runn1ng** [comment]
:sunglasses: great

File: test/fixtures/hdnode.json [hunk 7]
```diff
@@ -109,7 +109,7 @@
           "address": "19EuDJdgfRkwCmRzbzVBHZWQG9QNWhftbZ"
         },
         {
-          "description": "m/0/2147483647",
+          "description": "m/0/2147483647'",
```
Comment (dcousens) [inline test/fixtures/hdnode.json]: I might rename this to `path`

Comment (runn1ng) [inline test/fixtures/hdnode.json]: That would be great

File: src/hdnode.js [hunk 8]
```diff
@@ -291,6 +291,30 @@ HDNode.prototype.isNeutered = function () {
   return !(this.keyPair.d)
 }
 
+HDNode.prototype.derivePath = function (path) {
+  typeforce(types.Bip32Path, path)
+
+  var splitPath = path.split('/')
+  if (splitPath[0] === 'm') {
+    if (this.parentFingerprint) {
+      throw new Error('Not a master node')
```
Comment (dcousens) [inline src/hdnode.js]: Note,  I might change this error message in near future versions.  Consider unstable

File: src/hdnode.js [hunk 9]
```diff
@@ -291,6 +291,30 @@ HDNode.prototype.isNeutered = function () {
...
+  return splitPath.reduce(function (prevHd, indexStr) {
+    var index
+    if (indexStr.slice(-1) === "'") {
+      index = parseInt(indexStr.slice(0, -1), 10)
+      return prevHd.deriveHardened(index)
+    } else {
+      index = parseInt(indexStr, 10)
+      return prevHd.derive(index)
+    }
+  }, this)
```
Comment (dcousens) [inline src/hdnode.js]: nit: `this` unnecessary here

Comment (runn1ng) [inline src/hdnode.js]: Not sure, what would happen with empty paths

**Comment by runn1ng** [comment]
Thanks for merging, I kind of forgot about this PR :)

File: src/hdnode.js (hunk 9 above)
Comment (dcousens) [inline src/hdnode.js]: Oh woops, of course.  Ignore my comment.

**Pull request merged**